File: src/turtle_controller_py/turtle_controller_py/turtle_controller.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter

from rcl_interfaces.msg import SetParametersResult
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
from turtlesim.srv import SetPen

from turtle_controller_interfaces.srv import SwitchActivation


class TurtleControllerNode(Node):
    def __init__(self):
        super().__init__("turtle_controller")

        self.is_active = True

        self.declare_parameter("color_left", "green")
        self.declare_parameter("color_right", "red")
        self.declare_parameter("turtle_velocity", 1.0)

        self.color_left = self.get_parameter("color_left").value
        self.color_right = self.get_parameter("color_right").value
        self.turtle_velocity = self.get_parameter("turtle_velocity").value

        self.init_pen_color_settings()

        self.validate_params()

        self.pen_cur_color = self.color_left
        self.pen_request_pending = False
        self.pending_pen_color = None

        self.pose_sub = self.create_subscription(
            Pose, "/turtle1/pose", self.pose_callback, 10)
        self.cmd_vel_pub = self.create_publisher(
            Twist, "/turtle1/cmd_vel", 10)
        self.set_pen_client = self.create_client(
            SetPen, "/turtle1/set_pen")
        self.switch_activation_service = self.create_service(
            SwitchActivation, "switch_activation", self.switch_activation_service_callback)
        self.add_on_set_parameters_callback(self.parameters_callback)

        self.get_logger().info("Turtle Controller has been started.")

    def validate_velocity_param(self, velocity, errors) -> None:
        min_velocity = 0.0
        max_velocity = 3.0
        if not min_velocity < velocity <= max_velocity:
            errors.append(
                f"Invalid value for 'turtle_velocity': '{velocity}'.\n"
                f"Valid range is (0.0, 3.0]."
            )

    def validate_color_param(self, color, side, errors) -> None:
        if color not in self.pen_colors:
            errors.append(
                f"Invalid value for '{side}': '{color}'."
            )
 
    def validate_params(self) -> None:
        """Validate all configurable node parameters."""
        color_errors = []
        velocity_errors = []

        self.validate_color_param(self.color_left, "color_left", color_errors)
        self.validate_color_param(self.color_right, "color_right", color_errors)
        self.validate_velocity_param(self.turtle_velocity, velocity_errors)

        if color_errors:
            valid_colors = ", ".join(self.pen_colors.keys())
            color_errors.append(f"Valid colors are: {valid_colors}.")

        errors = color_errors + velocity_errors

        if errors:
            raise ValueError("\n".join(errors))

    def init_pen_color_settings(self) -> None:
        self.pen_colors = {
            "green" : (0, 255, 0),
            "red"   : (255, 0, 0),
            "yellow": (255, 255, 0),
            "purple": (128, 0, 128),
            "orange": (255, 165, 0),
            "cyan"  : (0, 255, 255)
        }

    def call_set_pen(self, color_name) -> None:
        """Send a request to change the turtle's pen color."""
        if self.pen_request_pending:
            # Keep only the most recently requested color while a SetPen
            # request is still in progress.
            self.pending_pen_color = color_name
            return
        
        while not self.set_pen_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().warning("Waiting for service...")

        r, g, b = self.pen_colors[color_name]
        request = SetPen.Request()
        request.r = r
        request.g = g
        request.b = b
        self.pen_request_pending = True

        future = self.set_pen_client.call_async(request)
        future.add_done_callback(
            lambda future: self.set_pen_callback(future, color_name))

    def pose_callback(self, pose: Pose) -> None:
        if not self.is_active:
            return

        screen_middle = 5.5
        cmd = Twist()

        if pose.x < screen_middle:
            cmd.linear.x = self.turtle_velocity
            cmd.angular.z = self.turtle_velocity
            if self.pen_cur_color == self.color_right:
                self.call_set_pen(self.color_left)
        else:
            cmd.linear.x = self.turtle_velocity * 2.0
            cmd.angular.z = self.turtle_velocity * 2.0
            if self.pen_cur_color == self.color_left:
                self.call_set_pen(self.color_right)

        self.cmd_vel_pub.publish(cmd)

    def set_pen_callback(self, future, color_name) -> None:
        future.result()

        self.pen_cur_color = color_name
        self.pen_request_pending = False

        self.get_logger().info(f"Pen color changed to {color_name}.")

        # Apply a color change that was requested while the previous
        # SetPen request was still pending.
        if self.pending_pen_color is not None:
            next_color = self.pending_pen_color
            self.pending_pen_color = None

            if next_color != self.pen_cur_color:
                self.call_set_pen(next_color) 

    def switch_activation_service_callback(self, request, response):
        if request.activate == self.is_active:
            response.success = False
            response.message = (
                "Turtle already activated."
                if self.is_active
                else "Turtle already deactivated."
            )
            return response

        self.is_active = request.activate
        response.success = True
        response.message = (
            "Turtle activated."
            if self.is_active
            else "Turtle deactivated."
        )

        return response

    def parameters_callback(
            self, 
            params: list[Parameter]
    ) -> SetParametersResult:
        """Validate and apply parameter changes at runtime."""
        color_errors = []
        velocity_errors = []

        for param in params:
            if param.name in {"color_left", "color_right"}:
                self.validate_color_param(param.value, param.name, color_errors)
            elif param.name == "turtle_velocity":
                self.validate_velocity_param(param.value, velocity_errors)

        if color_errors:
            valid_colors = ", ".join(self.pen_colors.keys())
            color_errors.append(f"Valid colors are: {valid_colors}.")

        errors = color_errors + velocity_errors

        if errors:
            return SetParametersResult(
                successful=False,
                reason="\n".join(errors)
            )

        for param in params:
            if param.name == "color_left":
                if self.pen_cur_color == self.color_left:
                    self.call_set_pen(param.value)
                self.color_left = param.value
            elif param.name == "color_right":
                if self.pen_cur_color == self.color_right:
                    self.call_set_pen(param.value)
                self.color_right = param.value
            elif param.name == "turtle_velocity":
                self.turtle_velocity = param.value

        return SetParametersResult(successful=True)


def main(args=None):
    rclpy.init(args=args)
    node = TurtleControllerNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/turtle_controller_cpp/src/turtle_controller.cpp
#include <cstdint>
#include <map>
#include <vector>
#include <string>
#include <stdexcept>

#include "rclcpp/rclcpp.hpp"
#include "rcl_interfaces/msg/set_parameters_result.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "turtlesim/msg/pose.hpp"
#include "turtlesim/srv/set_pen.hpp"
#include "turtle_controller_interfaces/srv/switch_activation.hpp"

using namespace std::placeholders;
using namespace std::chrono_literals;
using Pose = turtlesim::msg::Pose;
using Twist = geometry_msgs::msg::Twist;
using Pen = turtlesim::srv::SetPen;
using SwitchActivation = turtle_controller_interfaces::srv::SwitchActivation;

class TurtleControllerNode : public rclcpp::Node
{
public:

    TurtleControllerNode() : Node("turtle_controller")
    {
        is_active_ = true;

        this->declare_parameter("color_left", "green");
        this->declare_parameter("color_right", "red");
        this->declare_parameter("turtle_velocity", 1.0);

        color_left_ = this->get_parameter("color_left").as_string();
        color_right_ = this->get_parameter("color_right").as_string();
        turtle_velocity_ = this->get_parameter("turtle_velocity").as_double();

        init_pen_color_settings();

        validate_params();

        pen_cur_color_ = color_left_;
        pending_pen_color_ = "";
        pen_request_pending_ = false;

        pose_sub_ = this->create_subscription<Pose>(
            "/turtle1/pose",
            10,
            std::bind(&TurtleControllerNode::pose_callback, this, _1)
        );
        cmd_vel_pub_ = this->create_publisher<Twist>("/turtle1/cmd_vel", 10);
        set_pen_client_ = this->create_client<Pen>("/turtle1/set_pen");
        switch_activation_service_ = this->create_service<SwitchActivation>(
            "switch_activation",
            std::bind(&TurtleControllerNode::switch_activation_service_callback, this, _1, _2)
        );
        param_callback_handle_ = this->add_on_set_parameters_callback(
            std::bind(
                &TurtleControllerNode::parameters_callback,
                this,
                _1
            )
        );

        RCLCPP_INFO(this->get_logger(), "Turtle Controller has been started.");
    }
    
private:

    struct Color
    {
        uint8_t r;
        uint8_t g;
        uint8_t b;
    };

    void init_pen_color_settings()
    {   
        pen_colors_ = {
            {"green", {0, 255, 0}},
            {"red", {255, 0, 0}},
            {"yellow", {255, 255, 0}},
            {"purple", {128, 0, 128}},
            {"orange", {255, 165, 0}},
            {"cyan", {0, 255, 255}}
        };
    }

    void validate_color_param(
        const std::string color,
        const std::string side,
        std::vector<std::string>& errors)
    {
        if (pen_colors_.find(color) == pen_colors_.end()) {
            errors.push_back(
                "Invalid value for '" + side + "': '" + color + "'."
            );
        }
    }

    void validate_velocity_param(
        double velocity,
        std::vector<std::string>& errors)
    {
        const double min_velocity = 0.0;
        const double max_velocity = 3.0;

        if (!(min_velocity < velocity && velocity <= max_velocity)) {
            errors.push_back(
                "Invalid value for 'turtle_velocity': '" +
                std::to_string(velocity) +
                "'.\nValid range is (0.0, 3.0]."
            );
        }
    }

    /// Validate all configurable node parameters.
    void validate_params()
    {
        std::vector<std::string> color_errors;
        std::vector<std::string> velocity_errors;

        validate_color_param(color_left_, "color_left", color_errors);
        validate_color_param(color_right_, "color_right", color_errors);
        validate_velocity_param(turtle_velocity_, velocity_errors);

        if (!color_errors.empty()) {
            std::string valid_colors = "Valid colors are: ";
            bool first = true;

            for (const auto& entry : pen_colors_) {
                if (!first) {
                    valid_colors += ", ";
                }

                valid_colors += entry.first;
                first = false;
            }

            valid_colors += ".";
            color_errors.push_back(valid_colors);
        }

        std::vector<std::string> errors = color_errors;

        errors.insert(
            errors.end(),
            velocity_errors.begin(),
            velocity_errors.end()
        );

        if (!errors.empty()) {
            std::string message;

            for (std::size_t i = 0; i < errors.size(); ++i) {
                if (i > 0) {
                    message += "\n";
                }

                message += errors[i];
            }

            throw std::invalid_argument(message);
        }
    }

    /// Send a request to change the turtle's pen color.
    void call_set_pen(std::string color_name) 
    {
        if (pen_request_pending_) {
            // Keep only the most recently requested color.
            pending_pen_color_ = color_name;
            return;
        }

        while (!set_pen_client_->wait_for_service(1s)) {
            RCLCPP_WARN(this->get_logger(), "Waiting for the server...");
        }

        auto request = std::make_shared<Pen::Request>();
        const Color color = pen_colors_.at(color_name);
        request->r = color.r;
        request->g = color.g;
        request->b = color.b;

        pen_request_pending_ = true;

        set_pen_client_->async_send_request(
            request,
            // Capture the requested color so the confirmed state can be updated
            // when the asynchronous service call completes.
            [this, color_name](rclcpp::Client<Pen>::SharedFuture future)
            {
                set_pen_callback(future, color_name);
            }
        );
    }

    /// Porcess the result of a pen color change request.
    void set_pen_callback(rclcpp::Client<Pen>::SharedFuture future, std::string color_name)
    {
        future.get();
        pen_cur_color_ = color_name;
        pen_request_pending_ = false;
        RCLCPP_INFO(this->get_logger(), "Pen color has changed to %s.", pen_cur_color_.c_str());
        
        if (!pending_pen_color_.empty()) {
            std::string next_color = pending_pen_color_;
            pending_pen_color_.clear();

            if (next_color != pen_cur_color_) {
                call_set_pen(next_color);
            }
        }
    }

    void pose_callback(const Pose::SharedPtr pose)
    {
        if (!is_active_) {
            return;
        }

        const double screen_middle = 5.5;
        auto cmd = Twist();

        if (pose->x < screen_middle) {
            cmd.linear.x = turtle_velocity_;
            cmd.angular.z = turtle_velocity_;
            if (pen_cur_color_ == color_right_) {
                call_set_pen(color_left_);
            }
        } else {
            cmd.linear.x = turtle_velocity_ * 2.0;
            cmd.angular.z = turtle_velocity_ * 2.0;
            if (pen_cur_color_ == color_left_) {
                call_set_pen(color_right_);
            }
        }
        cmd_vel_pub_->publish(cmd);
    }

    void switch_activation_service_callback(
        const SwitchActivation::Request::SharedPtr request,
        const SwitchActivation::Response::SharedPtr response)
    {
        if (request->activate == is_active_) {
            response->success = false;
            response->message = is_active_
                ? "Turtle already activated."
                : "Turtle already deactivated.";
            return;
        }

        is_active_ = request->activate;
        response->success = true;
        response->message = is_active_
            ? "Turtle activated."
            : "Turtle deactivated.";
    }

    /// Validate and apply parameter changes at runtime.
    rcl_interfaces::msg::SetParametersResult parameters_callback(
        const std::vector<rclcpp::Parameter>& params)
    {
        std::vector<std::string> color_errors;
        std::vector<std::string> velocity_errors;

        for (const auto& param : params) {
            if (param.get_name() == "color_left" ||
                param.get_name() == "color_right"
            ) {
                validate_color_param(
                    param.as_string(), 
                    param.get_name(), 
                    color_errors
                );
            }
            else if (param.get_name() == "turtle_velocity") {
                validate_velocity_param(
                    param.as_double(),
                    velocity_errors
                );
            }
        }

        if (!color_errors.empty()) {
            std::string valid_colors = "Valid colors are: ";
            bool first = true;

            for (const auto& entry : pen_colors_) {
                if (!first) {
                    valid_colors += ", ";
                }

                valid_colors += entry.first;
                first = false;
            }

            valid_colors += ".";
            color_errors.push_back(valid_colors);
        }

        std::vector<std::string> errors = color_errors;
        
        errors.insert(
            errors.end(),
            velocity_errors.begin(),
            velocity_errors.end()
        );
        
        rcl_interfaces::msg::SetParametersResult result;

        if (!errors.empty()) {
            std::string message;

            for (std::size_t i = 0; i < errors.size(); ++i) {
                if (i > 0) {
                    message += "\n";
                }
                message += errors[i];
            }

            result.successful = false;
            result.reason = message;
            return result;
        }

        for (const auto& param : params) {
            if (param.get_name() == "color_left") {
                if (pen_cur_color_ == color_left_) {
                    call_set_pen(param.as_string());
                }
                color_left_ = param.as_string();
            }
            else if (param.get_name() == "color_right") {
                if (pen_cur_color_ == color_right_) {
                    call_set_pen(param.as_string());
                }
                color_right_ = param.as_string();
            }
            else if (param.get_name() == "turtle_velocity") {
                turtle_velocity_ = param.as_double();
            }
        }

        result.successful = true;
        return result;
    }

    std::string pen_cur_color_;
    std::string pending_pen_color_;
    std::map<std::string, Color> pen_colors_;
    std::string color_left_;
    std::string color_right_;
    bool is_active_;
    bool pen_request_pending_;
    double turtle_velocity_;

    rclcpp::Subscription<Pose>::SharedPtr pose_sub_;
    rclcpp::Publisher<Twist>::SharedPtr cmd_vel_pub_;
    rclcpp::Client<Pen>::SharedPtr set_pen_client_;
    rclcpp::Service<SwitchActivation>::SharedPtr switch_activation_service_;
    rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TurtleControllerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}